src/main/java/org/dependencytrack/notification/NotificationRouter.java

```diff
@@ -218,12 +220,50 @@ List<NotificationRule> resolveRules(final PublishContext ctx, final Notification
             pm.detachCopyAll(result);
             LOGGER.debug("Matched %d notification rules (%s)".formatted(result.size(), ctx));
 
+            final List<NotificationRule> severityFiltered = new ArrayList<>();
+            for (final NotificationRule rule : result) {
```

Comment:
I'd move "if (severities == null || severities.isEmpty())" here (before "for" block) and do severityFiltered=result; in this case, not to enumerate through all the results if severities are not set. Just to skip the whole following block